toolchain/check/testdata/impl/lookup/min_prelude/specialization.carbon

```diff
@@ -305,3 +305,110 @@ impl C(()) as Z where .X = C(()) {}
 fn F() {
   let a: C(()).(Z.X) = {} as C(());
 }
+
+// --- todo_fail_specialization_written_after_use_is_poisoned.carbon
+library "[[@TEST_NAME]]";
+
+interface Z(T:! type) {
+  let X:! type;
+}
+
+class C {}
+
+impl forall [T:! type] T as Z(T) where .X = () {}
+
+fn F() {
+  let a: C.(Z(C).X) = ();
+}
+
+// TODO: This impl changes the result of a previous lookup. The previous lookup
+// should have created a poison value that this trips and thus generates a
+// diagnostic.
+impl C as Z(C) where .X = C {}
+
+// --- fail_todo_final_specialization_before_generic_use_of_type_constant.carbon
+library "[[@TEST_NAME]]";
+
+interface Z(T:! type) {
+  let X:! type;
+}
+
+class C {}
+
+impl forall [T:! type] T as Z(T) where .X = () {}
+
+// This impl is effectively final, since it's all concrete.
+impl C as Z(C) where .X = C {}
+
+fn F[U:! type](T:! Z(U)) {
+  // TODO: The value of `.X` can be known to be `C` here.
```

Comment:
I don't see this -- the effectively final impl is too specific to apply here.

Reply:
Hm, yeah, making a specialization that's still generic was a bit trickier than I was thinking. Done, I think.